zrml/neo-swaps/src/lib.rs

```diff
@@ -105,6 +117,11 @@ mod pallet {
 
         type WeightInfo: WeightInfoZeitgeist;
 
+        /// The maximum allowed liquidity tree depth per pool. Each pool can support `2^(depth + 1)
+        /// - 1` liquidity providers. **Must** be less than 16.
```

Comment:
> What happens, if the 1025 LP wants to join?

The call will error with `LiquidityTreeFull`.

> Whats about the scenario that one entity creates a lot of LP accounts, each with the smallest liquidity provision possible (all joined)?

An attacker might attempt to execute a Denial of Service (DoS) attack. This attack comes in two flavors, both griefing-only vectors: preventing others from joining the tree and degrading its performance.

The first aspect of the attack, preventing others from joining the tree, would involve an attacker adding liquidity repeatedly to fill up all the positions in the tree. Our current structure counters this by increasing the required liquidity for each position by 1%. Therefore, to prevent others from joining the tree, the attacker would need to provide the $1.01^N \approx 26612$ fold of the initial required liquidity at the final position. Given a minimum liquidity of 10 DOT, this would amount to a substantial sum, well over a million dollars at the current market price. Moreover, if the attacker withdraws their liquidity, the abandoned nodes are reassigned, forcing the attacker to remain a liquidity provider to maintain the attack.

The second aspect of the attack, degrading the performance of the tree, is largely mitigated by the design of the liquidity tree. The operations of adding to or withdrawing from the tree have time complexity of $O(\log n)$, due to the binary tree structure. Therefore, regardless of the number of nodes in the tree, the time required to perform these operations increases logarithmically, not linearly. This limits the effectiveness of an attack aimed at degrading performance by clogging the tree with nodes.

In summary, although a determined attacker might attempt to clog the liquidity tree, the economic and computational challenges posed by such an attack make it unlikely to be successful or sustained.

> I currently see the MIN_RELATIVE_LP_POSITION_VALUE, so that's the relative minimum share of each liquidity provider against to the others?

Depends on when the two LPs whose stake you're comparing have joined the tree and what amounts all other LPs have brought. The last LP to join must have at least `MIN_RELATIVE_LP_POSITION_VALUE^1024` times the stake of the first LP.